pkg/router/template/router.go

```diff
@@ -311,31 +323,171 @@ func (r *templateRouter) readState() error {
 	return json.Unmarshal(data, &r.state)
 }
 
-// Commit applies the changes made to the router configuration - persists
+// Define a type for the pluggable commit function
+type CommitFunc func() error
+
+// This should ONLY be used for external testing hooks
+func (r *templateRouter) SetCommitFunc(commitFunc CommitFunc) {
+	r.commitFunc = commitFunc
+}
+
+// Commit calls the realCommit worker with a flag saying that it has
+// synced.  This should be called only when the router state is
+// consistent.  realCommit() below can be called on initial router
+// load to load the haproxy without a full state sync.
+func (r *templateRouter) Commit() {
+	r.realCommit(true)
+}
+
+// realCommit applies the changes made to the router configuration - persists
 // the state and refresh the backend. This is all done in the background
 // so that we can rate limit + coalesce multiple changes.
+//
+// The hasSynced argument specifies whether the state is fully synchronized.
+//
 // Note: If this is changed FakeCommit() in fake.go should also be updated
-func (r *templateRouter) Commit() {
+func (r *templateRouter) realCommit(hasSynced bool) {
 	r.lock.Lock()
 
-	if !r.synced {
-		glog.V(4).Infof("Router state synchronized for the first time")
-		r.synced = true
-		r.stateChanged = true
+	if hasSynced {
```

Comment:
I'm confused... if hasSynced = true that implies the state is fully synchronized, but then below r.synced can be false?  maybe there's a better name.

Reply:
If the user tells us we have synced, but our internal state doesn't yet reflect that... update the internal state and emit a message